### ISSUE TITLE
Missing main tag - important

### ISSUE THREAD

**Issue by rodolphebertozzo** [soflyy/breakdance-bugs#567]
You should include an automatically included `<main>` tag on every page and single article. Adapting a builder to meet the standards set by search engines in terms of semantic markup is important.

---

It's important to provide flexibility for `aside` tags, for example:
<img width="510" alt="image" src="https://github.com/soflyy/breakdance-bugs/assets/6425257/05ea0d0e-be18-4aac-a296-ace066be96c7">

**Comment by orhanbilikvar** [comment]
This is important

**Comment by JeusuDigital** [comment]
Not sure this would work as if you added an aside in the template it would wrap that also.
To have the main element there all you need to do is change the first section on your template files to be main element in advanced settings. And then for everything else you can set a catch all template with section set to main and then content area.

**Comment by rodolphebertozzo** [comment]
But no, it is a dynamically injected tag to enclose the main content as it is done on all web pages (it is placed after the header and stops before the.
This tag covers the entire content, whether it is an article or a page

**Comment by JeusuDigital** [comment]
I am aware of what the main tag is.
What I am saying is it would be difficult to create certain layouts if template files were all wrapped in it automatically.

Look here: https://www.w3.org/WAI/ARIA/apg/example-index/landmarks/complementary.html

Click show landmarks or view page source and you will see a use case.

**Comment by orhanbilikvar** [comment]
The use of the Aside section outside of the Main tag means that it is completely independent of the page content. This use may also be a need. But there is no point in discussing them when there is no Main tag. (GT)

**Comment by JeusuDigital** [comment]
But there is a main tag it is just manually set not automatic.
Create a template as per below, set it to everywhere and all your posts and pages will be wrapped in a main tag.

![screenshot](https://user-images.githubusercontent.com/29777022/208070547-cf98e8c6-28bf-4bbc-a239-a8eceb20dfd6.png)

**Comment by orhanbilikvar** [comment]
No, this only changes the label of a section. It is currently not possible to define a section within a section in BD.
We need something to wrap the sections. This way we can mark the content area as semantically.
Example;
<nav>...
<main>
--<article>...
--<section>...
--<section>...
--<aside>...
<footer>

**Comment by JeusuDigital** [comment]
I just created a post with only one template setup in Breakdance as above, below is the source, is that not what you are looking to do?

![chrome_qaMEYTeuFc](https://user-images.githubusercontent.com/29777022/208087092-1da9fdeb-ce32-4b75-8582-3de8c3abb16c.png)

**Comment by orhanbilikvar** [comment]
But this needs to be done in the Structure section in the BD editor. You cannot nest sections here. You nest divs, but they don't have the same functionality as sections. (GT)

**Comment by orhanbilikvar** [comment]
Being able to do it in the Structure section will make many functions possible. Normal operation should be like this.
Using custom templates or other solutions may be possible. But the goal is to develop the software in a way that naturally supports this. (GT)

**Comment by rodolphebertozzo** [comment]
@orhanbilikvar, you got it right, thanks for your help, now let's stop all unnecessary speculation and let the developers integrate it

**Comment by a-dubiel** [comment]
Planned

**Comment by Hakira-Shymuy** [comment]
Hello, im here to "remember" breakdance team about the Main tag

this issue was opened at November 24th of 2022, added to planned at april 13th of 2023.. something here got forgotten, come on guys,

**Comment by louisreingold** [comment]
We tried and it broke all sorts of backwards compatibility, styling, etc.

The workaround is as follows: create a template that applies to your pages, put a section in it, choose main, make the section full width, and add a Template Content Area to it.

**Comment by louisreingold** [comment]
Removing planned label, but not to say we won't come up with a better way to address this than the solution above, but I want to manage expectations

**Comment by Hakira-Shymuy** [comment]
> We tried and it broke all sorts of backwards compatibility, styling, etc.
> 
> The workaround is as follows: create a template that applies to your pages, put a section in it, choose main, make the section full width, and add a Template Content Area to it.

so basically we need a template for singles and 1 for each of the archives with a template area in it if I'm understanding you correctly?
For each type of page / archive / page view, we will need to apply a section to it
set full width, and padding 0 apply all

Thank you @louisreingold for a workaround

Btw, I would rather have this fixed sooner rather than later when we are on a version 1.8 or version 2 :( 

Maybe there could be a default template (like the one what we are applying at moment) by default? some "setting" "wrap in main tag" or maybe the "template area" could apply a wrapper (main)

Hope you guys find a solution for it

**Comment by rodolphebertozzo** [comment]
Do the best you can, but creating a page template with just one section tagged 'in hand' is hard to manage, especially as you can't insert sections within a section, and sections are designed from the outset to be adjusted as separate sections (alignment, spacing, flex or grid positioning of elements). And from an ergonomic and workflow point of view, it's still more meaningful to just insert sections automatically tagged as "section" with what you want inside (divs, columns), and dynamically generate the addition of the main attribute after the header and before the footer.

**Comment by Hakira-Shymuy** [comment]
> @louisreingold Do the best you can, but creating a page template with just one section tagged 'in hand' is hard to manage, especially as you can't insert sections within a section, and sections are designed from the outset to be adjusted as separate sections (alignment, spacing, flex or grid positioning of elements). And from an ergonomic and workflow point of view, it's still more meaningful to just insert sections automatically tagged as "section" with what you want inside (divs, columns), and dynamically generate the addition of the main attribute after the header and before the footer. I don't think the section element is suitable for designing an entire page template.

@rodolphebertozzo you cant create the "section as main tag" per page
Mainly because you do need to have 1 main tag per page, and does not make sense to create "per page" a section with main tag and then use divs inside

you create a "template" and apply to "pages" open the template, put a Section
Make full width
padding all to 0
change tag to main

inside that section insert a "template content area" element, done, save it

edit pages as normally you would with sections and so on

**Comment by orhanbilikvar** [comment]
It may be the solution for Pages only. You should look at posts and blogs.
It doesn't seem to cause a problem in the design.
(GT)
breakdance-no-template.php
/
...
outputHeadHtml();
echo $renderedHeader;
?>
<main>
<?php
echo $content;
?>
</main>
<footer>
<?php
echo $renderedFooter;

outputFootHtml();

?>
</footer>

![2023-07-20 10 02 55](https://github.com/soflyy/breakdance-bugs/assets/48365764/be39c441-d59b-4fe3-918e-a08b2b842228)

**Comment by orhanbilikvar** [comment]
For Posts and Blogs
breakdance-template.php

![2023-07-20 10 21 49](https://github.com/soflyy/breakdance-bugs/assets/48365764/4419b50e-63e4-4b8a-98f8-e50782f2d8cd)

**Comment by louisreingold** [comment]
Everyone, I need feedback before we ship this in 1.7: https://www.loom.com/share/54edfa760486402e9c2a03ab99ef4c6b

**Comment by louisreingold** [comment]
@rodolphebertozzo @orhanbilikvar @JeusuDigital @Hakira-Shymuy 

Everyone, I have what I think is a great solution, and better than anything else proposed in this thread:
 - it's really easy to implement (add a template, apply to `Pages`, add a `Template Content Area`)
 - it does not require nesting a template content area inside sections
 - it doesn't produce any extra HTML - just a main tag exactly where you want it
 - it retains total flexibility so that you can have separate `main` and `aside` tags for blog posts, etc.

Here's a demo: https://www.loom.com/share/54edfa760486402e9c2a03ab99ef4c6b

I need feedback before we ship this in 1.7.

**Comment by louisreingold** [comment]
I'd also like to note that unless I'm missing something, the solution proposed here achieves essentially the same result as the one mentioned above... and it's possible today. Unless there's any difference between `role="main"` and `<main>`... you can do what you want like this, in the current version of Breakdance... today (and since day 1).

<img width="683" alt="image" src="https://github.com/soflyy/breakdance-bugs/assets/6425257/97184389-6266-4bc9-8032-1410409bd6b7">

(https://breakdance.canny.io/features/p/a11y-add-main-element-as-wrapper-to-all-sections)

**Comment by louisreingold** [comment]
https://breakdance.canny.io/features/p/a11y-add-main-element-as-wrapper-to-all-sections

**Comment by louisreingold** [comment]
Can anyone comment where the "skip links" needs to gox? Inside the `<header>` is OK? Or not?

**Comment by orhanbilikvar** [comment]
+ looks good

**Comment by orhanbilikvar** [comment]
Main content should take priority during reading. When the user enters a page, he should have direct access to the main content. That's why I never used it. There may be special circumstances.

**Comment by JeusuDigital** [comment]
> @rodolphebertozzo @orhanbilikvar @JeusuDigital @Hakira-Shymuy
> 
> Everyone, I have what I think is a great solution, and better than anything else proposed in this thread:
> 
> * it's really easy to implement (add a template, apply to `Pages`, add a `Template Content Area`)
> * it does not require nesting a template content area inside sections
> * it doesn't produce any extra HTML - just a main tag exactly where you want it
> * it retains total flexibility so that you can have separate `main` and `aside` tags for blog posts, etc.
> 
> Here's a demo: https://www.loom.com/share/54edfa760486402e9c2a03ab99ef4c6b
> 
> I need feedback before we ship this in 1.7.

Would this be backward compatible? i.e not applying the main tag automatically to templates already created?
Otherwise anyone who has already added main tags and possibly got any code reliance on them will end up with broken accessibility and code.

**Comment by Hakira-Shymuy** [comment]
> @rodolphebertozzo @orhanbilikvar @JeusuDigital @Hakira-Shymuy
> 
> Everyone, I have what I think is a great solution, and better than anything else proposed in this thread:
> 
> * it's really easy to implement (add a template, apply to `Pages`, add a `Template Content Area`)
> * it does not require nesting a template content area inside sections
> * it doesn't produce any extra HTML - just a main tag exactly where you want it
> * it retains total flexibility so that you can have separate `main` and `aside` tags for blog posts, etc.
> 
> Here's a demo: https://www.loom.com/share/54edfa760486402e9c2a03ab99ef4c6b
> 
> I need feedback before we ship this in 1.7.

looks okay to me

**Comment by louisreingold** [comment]
> @louisreingold Would this be backward compatible? i.e not applying the main tag automatically to templates already created?

No, it would be a breaking change - the Template Content Area will have the `main` tag. So we'll have to tell everyone that if they use the Template Content Area... to change the tag from `main` if they don't want that.

I guess we could do it in a non-breaking way but then everyone who uses the Template Content Area would have to manually set the tag to `main`. Maybe that's better?

Although `main` as the default tag for Template Content Area seems like it's the right choice 99.999% of the time.

**Comment by orhanbilikvar** [comment]
> Can anyone comment where the "skip links" needs to go? Inside the `<header>` is OK? Or not?

Main

**Comment by JeusuDigital** [comment]
> > @louisreingold Would this be backward compatible? i.e not applying the main tag automatically to templates already created?
> 
> No, it would be a breaking change - the Template Content Area will have the `main` tag. So we'll have to tell everyone that if they use the Template Content Area... to change the tag from `main` if they don't want that.
> 
> I guess we could do it in a non-breaking way but then everyone who uses the Template Content Area would have to manually set the tag to `main`. Maybe that's better?
> 
> Although `main` as the default tag for Template Content Area seems like it's the right choice 99.999% of the time.

I understand that is can be unset, my question is if I have already created templates on a website when updating to this new version of Breakdance would it update any existing templates so they now have a main tag on the content area or would this only happen if I was to create new templates on that website or update any existing templates?

If it is the case that it does not change existing templates then that would be fine as it would mean that nobody who has already manually created main tags would need to go back and change all their templates if not maybe it would be better as a manual option to start and then integrated as standard at a later stage.

**Comment by rodolphebertozzo** [comment]
@louisreingold Indeed, this seems to me to be a good solution ;)

**Comment by dreadedhamish** [comment]
In the video there is a <header> tag which is a direct child of <body>- is this a pending change or something we should be seeing now? I'm using a regular header (not the header builder) and I've got a breakdance div in-between (body > div > header).

Which makes up part of my request - can we fill this out by adding <header> and <footer> into the mix as direct children of <body>?

**Comment by misslaidlaw** [comment]
Just wanted to add my two cents. 

I’ve seen the Skip Links added before the header but there is no harm adding it in the header as most people choose to either skip to main or the footer.  Skip links are focusable no matter where you put them because they are links. Putting them outside of the header might be a better option because sometimes people put hooks before the header for other reasons. (Woocommerce notifications, hello bars, etc) 

After much thought the idea of the Template Content Area being a main tag is good but only suitable for pages and posts. Not necessarily for CPT where fields are being used. 

From my understanding an aside can be in the main if the aside is related to the main. Which most times they are because they could hold related categories, etc.

**Comment by rodolphebertozzo** [comment]
Solved in Breadkance 1.7. Thanks @louisreingold !

**Comment by jared-leddy** [comment]
For anyone else coming to this thread, we wrote this script for our BD websites. It works really, really well. 

[https://gist.github.com/jared-leddy/fab9d531c6c7795954c6f4a210fd26fc](https://gist.github.com/jared-leddy/fab9d531c6c7795954c6f4a210fd26fc)
